crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,139 @@
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    RFC6979_SEED *rfc6979_seed = RAND_DRBG_get_callback_data(drbg);
+
+    *pout = (unsigned char *)rfc6979_seed->x;
+    return rfc6979_seed->rlen;
+}
+
+static size_t rfc6979_get_nonce(RAND_DRBG *drbg, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
+{
+    RFC6979_SEED *rfc6979_seed = RAND_DRBG_get_callback_data(drbg);
+
+    *pout = (unsigned char *)rfc6979_seed->h;
+    return rfc6979_seed->rlen;
+}
+
+static int bits2int(BIGNUM *out, int qlen,
+    const unsigned char *message, size_t message_len)
+{
+    if (BN_bin2bn(message, (int)message_len, out) == NULL)
+        return 0;
+    if ((int)message_len * 8 > qlen)
+        return BN_rshift(out, out, (int)message_len * 8 - qlen);
+    return 1;
+}
+
+static int int2octets(unsigned char *out, const BIGNUM *num, int rlen)
+{
+    return BN_bn2binpad(num, out, rlen);
+}
+
+static int bits2octets(unsigned char *out, const BIGNUM *range,
+    const unsigned char *message, size_t message_len, BN_CTX *ctx)
+{
+    BIGNUM *num = BN_new();
+
+    if (!bits2int(num, BN_num_bits(range), message, message_len)
+       || !BN_mod(num, num, range, ctx)
+       || !BN_bn2binpad(num, out, BN_num_bytes(range)))
+    {
+        BN_clear_free(num);
+        return 0;
+    }
+
+    BN_clear_free(num);
+
+    return 1;
+}
+
+int bn_generate_dsa_deterministic_nonce(BIGNUM *out, const BIGNUM *range,
+    const BIGNUM *priv, const unsigned char *message,
+    size_t message_len, int hash_type, BN_CTX *ctx)
+{
+    int ret = 0;
+    RAND_DRBG *drbg = NULL;
+    RFC6979_SEED rfc6979_seed = { NULL, NULL, 0 };
+    unsigned int drbg_flags = RAND_DRBG_FLAG_HMAC;
+    int rlen = 0, qlen = 0;
+    unsigned char *x = NULL, *h = NULL, *T = NULL;
+
+    if ((qlen = BN_num_bits(range)) < 0
+       || (rlen = BN_num_bytes(range)) < 0)
+        goto cleanup;
+
+    if ((x = (unsigned char *)OPENSSL_malloc(rlen)) == NULL
+       || (h = (unsigned char *)OPENSSL_malloc(rlen)) == NULL
+       || (T = (unsigned char *)OPENSSL_malloc(rlen)) == NULL)
+        goto cleanup;
+
+    if (!int2octets(x, priv, rlen)
+       || !bits2octets(h, range, message, message_len, ctx))
+        goto cleanup;
+
+    rfc6979_seed.x = x;
+    rfc6979_seed.h = h;
+    rfc6979_seed.rlen = rlen;
+
+    if ((drbg = RAND_DRBG_new(hash_type, drbg_flags, NULL)) == NULL
+       || !RAND_DRBG_set_callbacks(drbg, rfc6979_get_entropy, NULL, rfc6979_get_nonce, NULL)
+       || !RAND_DRBG_set_callback_data(drbg, &rfc6979_seed))
+        goto cleanup;
+
+    /* TODO(3.0): Rewrite the following two lines and include "../rand/rand_lcl.h" if drbg min_lengths can be set using methods */
+    drbg->min_entropylen = rlen;
+    drbg->min_noncelen = rlen;
+
+    if (!RAND_DRBG_instantiate(drbg, NULL, 0))
+        goto cleanup;
+
+    while (1)
+    {
```

Comment:
brace on previous line.  I prefer "for ( ; ; )" but that's just my opinion.  while is used 54 times, for is used 156 times FWIW.